src/libtego/source/context.cpp

```diff
@@ -0,0 +1,1030 @@
+        }
+        break;
+        case tego_proxy_type_socks5:
+        {
+            vm["Socks5Proxy"] = QString::fromStdString(fmt::format("{}:{}", config.proxy.address, config.proxy.port));
+            if (!config.proxy.username.empty())
+            {
+                vm["Socks5ProxyUsername"] = QString::fromStdString(config.proxy.username);
+            }
+            if (!config.proxy.password.empty())
+            {
+                vm["Socks5ProxyPassword"] = QString::fromStdString(config.proxy.password);
+            }
+        }
+        break;
+        case tego_proxy_type_https:
+        {
+            vm["HTTPSProxy"] = QString::fromStdString(fmt::format("{}:{}", config.proxy.address, config.proxy.port));
+            if (!config.proxy.username.empty() || !config.proxy.password.empty())
+            {
+                vm["HTTPSProxyAuthenticator"] = QString::fromStdString(fmt::format("{}:{}", config.proxy.username, config.proxy.password));
+            }
+        }
+        break;
+    }
+
+    // set firewall ports
+    if (config.allowedPorts.size() > 0)
+    {
+        std::stringstream ss;
+
+        auto it = config.allowedPorts.begin();
+
+        ss << fmt::format("*:{}", *it);
+        for(++it; it < config.allowedPorts.end(); ++it)
+        {
+            ss << fmt::format(", *:{}", *it);
+        }
+
+        vm["ReachableAddresses"] = QString::fromStdString(ss.str());
+    }
+
+    // set bridges
+    if (config.bridges.size() > 0)
+    {
+        QVariantList bridges;
+        for(const auto& currentBridge : config.bridges)
+        {
+            bridges.append(QString::fromStdString(currentBridge));
+        }
+        vm["Bridge"] = bridges;
+        vm["UseBridges"] = "1";
+    }
+
+    this->torControl->setConfiguration(vm);
+}
+
+void tego_context::save_tor_daemon_config()
+{
+    TEGO_THROW_IF_NULL(this->torControl);
+    logger::trace();
+    this->torControl->saveConfiguration();
+}
+
+void tego_context::set_host_user_state(tego_host_user_state_t state)
+{
+    if (state == hostUserState)
+    {
+        return;
+    }
+
+    this->hostUserState = state;
+    this->callback_registry_.emit_host_user_state_changed(state);
+}
+
+std::unique_ptr<tego_user_id_t> tego_context::get_host_user_id() const
+{
+    TEGO_THROW_IF_NULL(this->identityManager);
+    auto userIdentity = this->identityManager->identities().first();
+
+    auto hostname = userIdentity->hostname().toUtf8();
+    tego_v3_onion_service_id serviceId(hostname.data(), TEGO_V3_ONION_SERVICE_ID_LENGTH);
+
+    return std::make_unique<tego_user_id_t>(serviceId);
+}
+
+tego_host_user_state_t tego_context::get_host_user_state() const
+{
+    return this->hostUserState;
+}
+
+void tego_context::send_chat_request(
+    const tego_user_id_t* user,
+    const char* message,
+    size_t messageLength)
+{
+    TEGO_THROW_IF_NULL(this->identityManager);
+    auto userIdentity = this->identityManager->identities().first();
```

Comment:
since this is a list of pointers (`QList<class UserIdentity*> m_identities;`), should there be a check on whether `userIdentity` is `nullptr`?

Reply:
Technically yes, practically no, as the 'first' UserIdentity is created when the IdentityManager is created.